### DISCUSSION TITLE
Multi-threading with linear solver MUMPS and SPRAL

### DISCUSSION THREAD

**Discussion by federico-campanini** [casadi/casadi#3935] [Q&A]
Hello,

I am here again to ask for your help.

I am using CasADi 3.6.7 within Matlab to solve an optimization problem with Ipopt.
I tried both the linear solvers MUMPS and SPRAL.
Whit the default configuration, MUMPS is more efficient for my application.
Which is the default configuration of these linear solvers? Are they taking advantage of multi-threading?

I used the environment variable OMP_NUM_THREADS to define the number of threads used by OpenMP. I tried to set it to 4 and 8 without any change on the computational time.
It seems like this variable does not have any impact on the linear solvers.

Is there anyone that can explain which is the default configuration of these linear solvers within CasADi?
Are they already working on multiple threads or a special build of the tool is required for that?

Many thanks in advance for your help!

**Comment by jgillis** [comment]
SPRAL is built specifically with openmp disabled ( https://github.com/casadi/casadi/blob/main/CMakeLists.txt#L1374 ) because enabling it would make our binaries prone to crashes if one mixes them with binaries featuring other openmp versions.
Mumps does parallellization without openmp.

In principle, Matlab is a controlled environment where ambiguities about openmp can be avoided and openmp could be made to work.
In practice, I don't have the budget to offer solid support for this.

You can probably create a fork with openmp enabled and have the binaries created by github actions.

**Reply by federicoc-wae** [reply]
Thanks a lot Joris for your answer. This explains the default configuration of MUMPS and SPRAL.

I will try to configure SPRAL with openmp and share the results. There is also a GPU option I would like to try.

Many Thanks!
